### DISCUSSION TITLE
Running libGL applications

### DISCUSSION THREAD

**Discussion by bobbbay** [nix-community/NixOS-WSL#88] [Q&A] [answered]
Hi all, I'm developing a project that depends on libGL to launch (using Rust's `egui`, if it helps). However, when I try to run it, I receive:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: CreationErrors([NotSupported("lacking either libGL or libEGL so could not fallback to other"), NotSupported("both libGL and libEGL are not present")])', /home/bob/.cargo/registry/src/github.com-1ecc6299db9ec823/egui_glow-0.17.0/src/epi_backend.rs:29:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I have a working X server, and can run other GUI applications. I'm *really* unfamiliar with the workings of GUI under-the-hood, which is why I have found no way to resolve this. I have tried adding some (confusingly named) GL libraries to my shell to no avail.

I'm writing this here because it feels less like a bug and more like something I'm doing wrong - I just don't know what it is :).

**Comment by nzbr** [comment]
glxgears works on Ubuntu but not on NixOS for me - so there might actually be a bug related to graphics acceleration in NixOS-WSL

**Comment by bobbbay** [comment]
Should I open a bug report, instead?

**Comment by shayne** [comment, marked as answer]
I ran into something similar a while back and solved my issue by adding: `hardware.opengl.enable = true;`

**Reply by nzbr** [reply]
This worked for me. Can you test that @bobbbay?

**Reply by bobbbay** [reply]
With the following extra configuration:

```nix
  hardware.opengl.enable = true;
  hardware.opengl.extraPackages = with pkgs; [ mesa.drivers ];
  hardware.opengl.driSupport32Bit = true;
```

AND running it under [NixGL](https://github.com/guibou/nixGL), I can run `glxinfo` :). But, running the project results in the following panic:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: CreationErrors([OsError("Could not create EGL display object"), OsError("Couldn't find any available vsync extension")])', /home/bob/.cargo/registry/src/github.com-1ecc6299db9ec823/egui_glow-0.17.0/src/epi_backend.rs:29:14
```

By the way, if I don't use NixGL, I get the following error from `glxinfo`:

```
name of display: 172.28.0.1:0
Error: couldn't find RGB GLX visual or fbconfig
```

And the following from the project:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: CreationErrors([NotSupported("lacking either libGL or libEGL so could not fallback to other"), NotSupported("both libGL and libEGL are not present")])', /home/bob/.cargo/registry/src/github.com-1ecc6299db9ec823/egui_glow-0.17.0/src/epi_backend.rs:29:14
```

**Reply by nzbr** [reply]
> Error: couldn't find RGB GLX visual or fbconfig
That's exactly the error i got before setting `hardware.opengl.enable = true;`, but now that it's enabled, i can run glxinfo and glxgears just fine. Can you run `glxinfo` in another WSL distro like Ubuntu?

My host has a Intel Iris Xe GPU, that's probably a factor

**Reply by bobbbay** [reply]
I definitely added it and rebuilt, but it doesn't help. I used to use Ubuntu, and that ran glxinfo/glxgears (as well as other GUI programs) just fine. To me, it's moreso interesting that NixGL solves this, but `hardware.opengl.*` don't. I have an integrated Intel UHD Graphics 620 (or something similar), too.

**Reply by nzbr** [reply]
Just to rule out your config being the problem, would you mind trying [this tarball](https://s3.eu-central-1.wasabisys.com/nzbr/nixos-wsl/nixos-wsl-installer-glxinfo.tar.gz)? 
I imported it and was able to start glxinfo just fine

SHA256: `3E2817B032BDED33396B9E28AD507D447AB790D34C9F46EF09F647D727BE48E4`

It's built from main and this config:
```nix
{ lib, pkgs, config, modulesPath, ... }:

with lib;
let
  nixos-wsl = import ./default.nix;
in
{
  imports = [
    "${modulesPath}/profiles/minimal.nix"

    nixos-wsl.nixosModules.wsl
  ];

  wsl = {
    enable = true;
    automountPath = "/mnt";
    defaultUser = "nixos";
    startMenuLaunchers = true;

    interop.register = false;
  };

  hardware.opengl.enable = true;

  environment.systemPackages = with pkgs; [
    glxinfo
  ];

  # Enable nix flakes
  nix.package = pkgs.nixFlakes;
  nix.extraOptions = ''
    experimental-features = nix-command flakes
  '';
}
```

**Reply by bobbbay** [reply]
Thank you, I really appreciate your help. I installed the tarball, set the DISPLAY with:

```shell
export DISPLAY=$(ip route list default | awk '{print $3}'):0
export LIBGL_ALWAYS_INDIRECT=1
```

And still get the error :).

```
[nixos@nixos:/]$ glxinfo
name of display: 172.31.208.1:0
Error: couldn't find RGB GLX visual or fbconfig


[nixos@nixos:/]$ glxgears
Error: couldn't get an RGB, Double-buffered visual

```

However, `nix run nixpkgs#xorg.xclock` works just fine.

**Reply by bobbbay** [reply]
I wonder if it has anything to do with the following option on my (Windows) X server:

![image](https://user-images.githubusercontent.com/49123594/166118574-9519d15c-8580-41fe-8353-e922417af05e.png)

Maybe OpenGL doesn't work because VcXsrv wants to use the Windows-native version instead? But, I just launched it with this option disabled, and get the same error.

**Reply by nzbr** [reply]
I was assuming that you are using WSLg. I didn't have a good experience using VcXsrv or Xming, so I used X410 before WSLg was a thing

That it works in Ubuntu means that there has to be _some_ way to get this to work, but unfortunately I have no Idea what that may be. 
Is using WSLg an option for you?

**Reply by bobbbay** [reply]
Same thing. Now, it gets weirder: `glxinfo` looks happy, and spews a bunch of output about my graphics (I assume). `glxgears`, however, is strange: it outputs on the cli, but no window opens/there is no visual. Last time I checked, it should open up two gears spinning quickly. The output I get is the following:

```
[nixos@NotYourLaptop:/]$ glxgears
2830 frames in 5.0 seconds = 565.870 FPS
2908 frames in 5.0 seconds = 581.525 FPS
3329 frames in 5.0 seconds = 665.776 FPS
3508 frames in 5.0 seconds = 701.523 FPS
3300 frames in 5.0 seconds = 659.997 FPS
3580 frames in 5.0 seconds = 715.875 FPS
3648 frames in 5.0 seconds = 729.592 FPS
3583 frames in 5.0 seconds = 716.479 FPS
3623 frames in 5.0 seconds = 724.501 FPS
3472 frames in 5.0 seconds = 694.234 FPS
3587 frames in 5.0 seconds = 717.256 FPS
2596 frames in 5.0 seconds = 519.089 FPS
2916 frames in 5.0 seconds = 583.097 FPS
3272 frames in 5.0 seconds = 654.378 FPS
3480 frames in 5.0 seconds = 695.864 FPS
3484 frames in 5.0 seconds = 696.606 FPS
3472 frames in 5.0 seconds = 694.295 FPS
3430 frames in 5.0 seconds = 685.858 FPS
...
```

Which I imagine is regular output. My laptop fans start going too, so I have a feeling that the graphic acceleration is really being used, I just can't see it (?). Again, not an expert in that area.

**Reply by nzbr** [reply]
WSLg can be really slow to showing the first window for some reason. Things that _can_ help are
- just waiting (in bad cases even multiple minutes)
- Stopping all instances of the RDP client (`mstsc.exe`) either in Task Manager or with `taskkill /f /im mstsc.exe`
- Rebooting Windows
I have experienced that on other distros as well, so there's probably nothing from the Linux side that can be done

**Reply by bobbbay** [reply]
Yes, it seems the issue I'm facing is an issue with WSLg (that others have had). Thank you for getting me this far, and I think it's safe to say that if this issue is resolved, libGL on NixOS will be fine. So, I'll close this discussion until further notice 🙂. Thanks again!
